keras_nlp/models/gpt2/gpt2_causal_lm.py

```diff
@@ -0,0 +1,333 @@
+    ds = tf.data.Dataset.from_tensor_slices(features)
+
+    # Create a `GPT2CausalLM` and fit your data.
+    gpt2_lm = keras_nlp.models.GPT2CausalLM.from_preset(
+        "gpt2_base_en",
+        preprocessor=preprocessor,
+    )
+    gpt2_lm.compile(
+        loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+    )
+    gpt2_lm.fit(x=features, batch_size=2)
+    ```
+
+    # Use `generate()` method to generate text.
+    ```python
+    gpt2_lm = keras_nlp.models.GPT2CausalLM.from_preset("gpt2_base_en")
+    gpt2_lm.generate("I want to say", max_length=30)
+
+    # Generate with batched prompts.
+    gpt2_lm.generate(["This is a", "Where are you"], max_length=30)
+    ```
+
+    """
+
+    def __init__(self, backbone, preprocessor=None, **kwargs):
+        inputs = backbone.input
+        x = backbone(inputs)
+        embedding_layer = backbone.get_layer("token_embedding")
+        embedding_map_layer = EmbeddingMapping(embedding_layer)
+        outputs = embedding_map_layer(x)
+
+        # Instantiate using Functional API Model constructor
+        super().__init__(
+            inputs=inputs,
+            outputs=outputs,
+            include_preprocessing=preprocessor is not None,
+            **kwargs,
+        )
+
+        self._backbone = backbone
+        self._preprocessor = preprocessor
+
+    def preprocess_samples(self, x, y=None, sample_weight=None):
+        return self.preprocessor(x, y=y, sample_weight=sample_weight)
+
+    @property
+    def backbone(self):
+        """The associated `keras_nlp.models.GPT2Backbone`."""
+        return self._backbone
+
+    @property
+    def preprocessor(self):
+        """A `keras_nlp.models.GPT2CausalLMPreprocessor` for preprocessing."""
+        return self._preprocessor
+
+    @classproperty
+    def presets(cls):
+        return copy.deepcopy(backbone_presets)
+
+    @classmethod
+    def from_preset(
+        cls,
+        preset,
+        load_weights=True,
+        **kwargs,
+    ):
+        if "preprocessor" not in kwargs:
+            kwargs["preprocessor"] = GPT2CausalLMPreprocessor.from_preset(
+                preset
+            )
+
+        # Check if preset is backbone-only model.
+        if preset in GPT2Backbone.presets:
+            backbone = GPT2Backbone.from_preset(preset, load_weights)
+            return cls(backbone, **kwargs)
+
+        # Otherwise must be one of class presets.
+        # Currently no classifier-level presets, so we raise ValueError.
+        if preset not in cls.presets:
+            raise ValueError(
+                "`preset` must be one of "
+                f"""{", ".join(cls.presets)}. Received: {preset}."""
+            )
+
+    def _get_token_probability(self, prompt, mask):
+        model_inputs = {
+            "token_ids": prompt,
+            "padding_mask": mask,
+        }
+        return self(model_inputs)
+
+    def generate(
+        self,
+        prompt,
+        max_length,
+        end_token="<|endoftext|>",
```

Comment:
We generally keep track of these on the base class tokenizer (which you could reference here). `self.preprocessor.tokenizer.end_token` and `self.preprocessor.tokenizer.end_token_id`.

Reply:
Yea I am aware of that, I am leaving this as a settable field in case users' dataset has a special end_token.

Reply:
This is an interesting question. Do we want the tokenizer to be the thing that "owns" the knowledge of special tokens? That is what we have been doing up to this point. There is zero way to configure the cls token, sep token for Bert, Roberta currently for example, and if we did, we might best expose this in the tokenizer classes themselves.

Why not be defensive for now and not expose this (read it hard coded from the tokenizer). If people ask for it, we can figure it out then?

This is setting a precedent I'm not sure we want to follow--"leaking" our preprocessing knowledge across our whole API surface. And it seems easy to just remove for now without breaking any workflows. Why not scope this down?

Reply:
SGTM!